crates/wysiwyg/src/composer_model/new_lines.rs

```diff
@@ -0,0 +1,423 @@
+            ]);
+            self.state.advance_selection();
+            return self.create_update_replace_all();
+        }
+
+        // If the selection covered several characters, remove them first
+        if range.is_selection() {
+            self.do_replace_text(S::default());
+        }
+
+        let block_location = range.deepest_block_node(None).expect(
+            "No block node selected (at least the root one should be here)",
+        );
+        let block_handle = block_location.node_handle.clone();
+
+        // TODO: what if a block node was removed and the next one has the same type?
+        // If we the block node was removed, just insert an empty paragraph
+        if !self.state.dom.contains(&block_handle)
+            || self.state.dom.lookup_node(&block_handle).kind()
+                != block_location.kind
+        {
+            let mut insert_at = block_handle.clone();
+            loop {
+                let has_prev = insert_at.index_in_parent() > 0;
+                if self.state.dom.contains(&insert_at)
+                    || (has_prev
+                        && self.state.dom.contains(&insert_at.prev_sibling()))
+                {
+                    break;
+                }
+                if insert_at.has_parent() {
+                    insert_at = insert_at.parent_handle();
+                } else {
+                    break;
+                }
+            }
+            let paragraph = DomNode::new_paragraph(Vec::new());
+            if insert_at.is_root() {
+                self.state.dom.append_at_end_of_document(paragraph);
+            } else {
+                self.state.dom.insert_at(&block_handle, paragraph);
+            }
+            return self.create_update_replace_all();
+        }
+
+        let first_leaf = range.leaves().next();
+        match block_location.kind {
+            Paragraph => {
+                let ancestor_block_location =
+                    range.deepest_block_node(Some(block_handle.clone()));
+                if let Some(ancestor_block_location) = ancestor_block_location {
+                    if ancestor_block_location.kind != Generic {
+                        if block_location.is_empty() {
+                            self.do_new_line_in_block_node(
+                                block_location,
+                                ancestor_block_location,
+                            );
+                        } else {
+                            self.do_new_line_in_paragraph(
+                                first_leaf,
+                                block_location,
+                            );
+                        }
+                    } else {
+                        self.do_new_line_in_paragraph(
+                            first_leaf,
+                            block_location,
+                        );
+                    }
+                } else {
+                    self.do_new_line_in_paragraph(first_leaf, block_location);
```

Comment:
I see how we can merge the 2 nested `do_new_line_in_paragraph` inside the `if let` check, but I don't see how we can merge them also with the external `else` clause.

Reply:
Oh yeah I guess the only way to do something like this would be to extract to a function and return when the action is done. We could do it in a match arm too I guess, but I'm sure everyone will hate something like this (😆 ): 

```
if let ... {
    if ... {
        do_something();
        break;
    }
}

do_default();
```

I guess repeating this only twice is decent enough